Add a helper command for one_gadget

lebr0nli opened the pull request:
This PR adds a new command for finding valid [one_gadget](https://github.com/david942j/one_gadget) for the current context and makes it easier to debug when the constraints are unsat.

----

Here are some example usages:

After we control pc, if we get some luck, we might be able to find a satisfiable gadget via command: `onegadget`:
<img width="969" alt="image" src="https://github.com/pwndbg/pwndbg/assets/61896187/805ee598-7f0e-4db0-aabf-cc653e39f55f">

If there's no satisfiable one_gadget for the current context, we can run with `--show-unsat` to find out what constraint is unsatisfiable: 
<img width="786" alt="image" src="https://github.com/pwndbg/pwndbg/assets/61896187/bdd0ee29-15bd-4590-b882-aee824cf4418">

If we want to know more information behind the unsatisfiable gadgets, we can run with `--verbose` or `-v`:
<img width="748" alt="image" src="https://github.com/pwndbg/pwndbg/assets/61896187/622656f0-0d68-41a1-aade-22ea225db2f5">

lebr0nli committed:
Add a helper command to find valid one_gadget for current context

lebr0nli committed:
Refactor the function for getting section address

disconnect3d commented:
Any chance we can display the condition and SAT/UNSAT in two columns? Or somehow that it will be more readable? I feel like the output is bloated with colors and maybe not so readable now.

We could also rename the command to just `onegadget` too 🤔

lebr0nli commented:
> Any chance we can display the condition and SAT/UNSAT in two columns? Or somehow that it will be more readable? I feel like the output is bloated with colors and maybe not so readable now.

Yeah, I also feel that the output is a little bit messy at the moment.
I'll attempt to organize it into a table format with columns to see if that offers a cleaner presentation.

> We could also rename the command to just `onegadget` too 🤔

Sure!

lebr0nli committed:
Rename the command to onegadget for more convenient typing

lebr0nli commented:
Regarding the output format, how about something like this?
<img width="1026" alt="SCR-20240219-mdcn" src="https://github.com/pwndbg/pwndbg/assets/61896187/918874aa-3ade-421a-92d4-8fcf81d99be3">

^ I think this looks much better(?), but it sacrifices the ability to display the detailed information behind its UNSAT status.
This results that I still need to check each UNSAT constraint manually to find out how to make it SAT when debugging.

Should we use the tabulate output as default, but add an `-v` option to show the original messy but detailed output?
Or any other ideas?

disconnect3d commented:
> ^ I think this looks much better(?), but it sacrifices the ability to display the detailed information behind its UNSAT status. This results that I still need to check each UNSAT constraint manually to find out how to make it SAT when debugging.
> 
> Should we use the tabulate output as default, but add an `-v` option to show the original messy but detailed output? Or any other ideas?

This looks good

lebr0nli committed:
Make the output format cleaner

lebr0nli commented:
> This looks good

Ok!
I'll update the output format to the tabulate one.

lebr0nli committed:
Add a simple cache mechanism for the one_gadget output

lebr0nli committed:
Update the warning message

lebr0nli committed:
Use MD5 instead of BLAKE2 for computing the file hash

I thought that BLAKE2 was faster than MD5, but it doesn't seem correct here somehow (probably because of the implementation of Python!?)

Here's the script I used for benchmarking:
```python
import hashlib
import timeit

def compute_file_hash_1() -> str:
    h = hashlib.blake2b()
    with open("/lib/x86_64-linux-gnu/libc.so.6", "rb") as f:
        h.update(f.read())
    return h.hexdigest()

def compute_file_hash_2() -> str:
    h = hashlib.md5()
    with open("/lib/x86_64-linux-gnu/libc.so.6", "rb") as f:
        h.update(f.read())
    return h.hexdigest()

print(timeit.timeit(compute_file_hash_1, number=1000))
print(timeit.timeit(compute_file_hash_2, number=1000))
```

I executed the above script on various machines, and the results seem to show that MD5 outperforms BLAKE2 in this scenario. (On my x86 VM running through QEMU on my M1 MacBook, BLAKE2 even takes almost twice as long as MD5.)

lebr0nli committed:
Add the tests for `onegadget` command

lebr0nli committed:
Fix lint issue

lebr0nli committed:
Try to cover more code

lebr0nli committed:
Fix lint issue

lebr0nli committed:
Fix illogical tests

anthraxx commented on pwndbg/commands/one_gadget.py:
```diff
@@ -0,0 +1,50 @@
+
+Examples:
+    one_gadget
+    one_gadget --show-unsat
```
The command has been renamed to `onegadget`

lebr0nli replied:
Ahh, yes, thanks, I'll fix it!

anthraxx commented on pwndbg/commands/one_gadget.py:
```diff
@@ -0,0 +1,50 @@
+
+parser = argparse.ArgumentParser(
+    formatter_class=argparse.RawTextHelpFormatter,
+    description="""Show one_gadget
```
The command has been renamed to `onegadget`

disconnect3d replied:
@lebr0nli perhaps also rename the file to `onegadget.py`

lebr0nli replied:
Sure, I'll rename it

lebr0nli committed:
Rename one_gadget to onegadget

lebr0nli committed:
Use `pwndbg.lib.tempfile.cachedir` for `onegadget`

lebr0nli committed:
Call `pwndbg.lib.tempfile.cachedir` only once

disconnect3d commented on pwndbg/gdblib/onegadget.py:
```diff
@@ -0,0 +1,595 @@
+VALID_ENVP_PATTERN = re.compile(r"(.+) is a valid envp")
+IS_ALIGNED_PATTERN = re.compile(r"(.+) & 0xf == (\d+)")
+IS_GOT_ADDRESS_PATTERN = re.compile(r"(.+) is the GOT address of libc")
+CAST_PATTERN = re.compile(r"^\([s|u]\d+\)")
```
Fwiw all this `re.compile` etc. will slow down Pwndbg load time, but it seems negligible:
```
pwndbg> !cat a.py
import re
import time

a = time.time()
ADDRESS_WRITABLE_PATTERN = re.compile(r"address(?:es)? (.*) (?:is|are) writable")
WRITABLE_COLON_PATTERN = re.compile(r"writable: (.*)")
EQUAL_NULL_PATTERN = re.compile(r"(.*) == NULL")
VALID_POSIX_SPAWN_FILE_ACTIONS_PATTERN = re.compile(r"(.*) <= 0")
VALID_ARGV_PATTERN = re.compile(r"(.+) is a valid argv")
VALID_ENVP_PATTERN = re.compile(r"(.+) is a valid envp")
IS_ALIGNED_PATTERN = re.compile(r"(.+) & 0xf == (\d+)")
IS_GOT_ADDRESS_PATTERN = re.compile(r"(.+) is the GOT address of libc")
CAST_PATTERN = re.compile(r"^\([s|u]\d+\)")
b = time.time()

elapsed = b - a
print(elapsed)
pwndbg> source a.py
0.0009448528289794922
```
